app/src/main/java/com/hedvig/app/feature/offer/OfferViewModel.kt

```diff
@@ -90,33 +103,49 @@ class OfferViewModelImpl(
                     quoteIds = idsResult.ids
                     idsResult
                         .data
-                        .map(::toViewState)
-                        .onEach(_viewState::postValue)
-                        .catch { _viewState.postValue(ViewState.Error.GeneralError(it.message)) }
+                        .onEach { response ->
+                            when (val result = toDataOrError(response)) {
+                                is Either.Left -> {
+                                    _viewState.postValue(toViewState(result.value))
+                                }
+                                is Either.Right -> {
+                                    _events.tryEmit(result.value)
+                                }
+                            }
+                        }
+                        .catch {
+                            _events.tryEmit(Event.Error.GeneralError(it.message))
+                        }
                         .launchIn(this)
                 }
                 GetQuotesUseCase.Result.Error -> {
-                    _viewState.postValue(ViewState.Error.GeneralError(""))
+                    _events.tryEmit(Event.Error.GeneralError(""))
                 }
             }
         }
     }
 
-    private fun toViewState(response: Response<OfferQuery.Data>): ViewState {
-        return response.errors?.let {
-            ViewState.Error.GeneralError(it.firstOrNull()?.message)
-        } ?: response.data?.let { data ->
-            if (data.contracts.isNotEmpty()) {
-                ViewState.HasContracts
-            } else {
-                val topOfferItems = OfferItemsBuilder.createTopOfferItems(data)
-                val perilItems = OfferItemsBuilder.createPerilItems(data.quoteBundle.quotes)
-                val insurableLimitsItems = OfferItemsBuilder.createInsurableLimits(data.quoteBundle.quotes)
-                val documentItems = OfferItemsBuilder.createDocumentItems(data.quoteBundle.quotes)
-                val bottomOfferItems = OfferItemsBuilder.createBottomOfferItems()
-                ViewState.OfferItems(topOfferItems, perilItems, documentItems, insurableLimitsItems, bottomOfferItems)
-            }
-        } ?: ViewState.Error.EmptyResponse
+    private fun toDataOrError(response: Response<OfferQuery.Data>): Either<OfferQuery.Data, Event> {
```

Comment:
I think we should try to move this down to `OfferRepository`, and return som kind of `OfferResult` instead of ``Response<OfferQuery.Data>` from `fun offer(ids: List<String>)`. This will be in line with not exposing the network layer to the view layer. Something like this (in `OfferRepository`):

```
fun offer(ids: List<String>) = apolloClient
        .query(offerQuery(ids))
        .watcher()
        .toFlow()
        .map(::toDataOrError) 
```     

What do you think?

Reply:
I see your point.

As the code is written now, the `toDataOrError`-function returns `Either<OfferQuery.Data, OfferViewModel.Event>`. I guess it would not make sense to expose the viewmodel-layer to the network layer by continuing to have the flow return `OfferViewModel.Event` in its either-return. Would it be overkill to introduce one more error-type and map from it to an `Event` at the `ViewModel`-layer? 🤔